lib/descriptor/common.js

```diff
@@ -0,0 +1,231 @@
+/*!
+ * common.js - common functions for descriptor
+ * Copyright (c) 2023, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const common = exports;
+
+/**
+ * Types of script expressions in descriptors
+ * @see https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#reference
+ * @const {Object}
+ */
+
+common.types = {
+  PK: 'pk',
+  PKH: 'pkh',
+  WPKH: 'wpkh',
+  SH: 'sh',
+  WSH: 'wsh',
+  COMBO: 'combo',
+  ADDR: 'addr',
+  MULTI: 'multi',
+  SORTEDMULTI: 'sortedmulti',
+  RAW: 'raw'
+};
+
+/**
+ * parse script context for descriptor
+ * @const {Object}
+ */
+
+common.scriptContext = {
+  TOP: 'TOP',
+  P2SH: 'P2SH',
+  P2WPKH: 'P2WPKH',
+  P2WSH: 'P2WSH'
+};
+
+/**
+ * Allowed characters in descriptor expressions for checksum to work.
+ * @see https://github.com/bitcoin/bips/blob/master/bip-0380.mediawiki#character-set
+ * @const {String}
+ */
+
+const INPUT_CHARSET = '0123456789()[],\'/*abcdefgh@:$%{}IJKLMNOPQRSTUVWXYZ' +
+                       '&+-.;<=>?!^_|~ijklmnopqrstuvwxyzABCDEFGH`#"\\ ';
+
+ /**
+ * Checksum character set.
+ * The checksum itself uses the same character set as bech32
+ * @const {String}
+ */
+
+const CHECKSUM_CHARSET = 'qpzry9x8gf2tvdw0s3jn54khce6mua7l';
+
+/**
+ * Test whether this descriptor string starts with particular script type
+ * ('pk', 'pkh', 'wpkh' etc.)
+ * @param {String} scriptType script type
+ * @param {String} desc descriptor string
+ * @returns {Boolean}
+ */
+
+common.isType = function isType(scriptType, desc) {
+  assert(typeof scriptType === 'string');
+  assert(typeof desc === 'string');
+
+  /**
+   * First check whether descriptor string length is atleast greater than
+   * scriptType length + 2 (for brackets)
+   * Then check character at scriptType length is '(' and last character is ')'
+   * in descriptor string
+   * Then check whether descriptor string starts with scriptType
+   */
+
+  return desc.length >= scriptType.length + 2
+         && desc[scriptType.length] === '('
+         && desc[desc.length - 1] === ')'
+         && scriptType === desc.substring(0, scriptType.length);
+};
+
+/**
+ * Get the top level script expression of the descriptor.
+ * @param {String} desc descriptor string
+ * @returns {String} script expression of descriptor
+ */
+
+common.getType = function getType(desc) {
+  return desc.substring(0, desc.indexOf('('));
+};
+
+/**
+ * Strip script expression string from descriptor
+ * @param {String} scriptType type of script expression
+ * @param {String} desc descriptor string
+ * @returns {String} descriptor string without script expression
+ */
+
+common.strip = function strip(scriptType, desc) {
+  desc = desc.substring(scriptType.length + 1, desc.length - 1);
```

Comment:
can't you just slice at the first `(` ? And then you won't need the caller to pass in the type every time?